safehttp/plugins/cors/cors.go

```diff
@@ -0,0 +1,214 @@
+// accordingly.
+func (it *Interceptor) Before(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.Result {
+	origin := r.Header.Get("Origin")
+	if origin != "" && !it.AllowedOrigins[origin] {
+		return w.ClientError(safehttp.StatusForbidden)
+	}
+	h := w.Header()
+	allowOrigin, err := h.Claim("Access-Control-Allow-Origin")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+	vary, err := h.Claim("Vary")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	allowCredentials, err := h.Claim("Access-Control-Allow-Credentials")
+	if err != nil {
+		return w.ServerError(safehttp.StatusInternalServerError)
+	}
+
+	var status safehttp.StatusCode
+	if r.Method() == safehttp.MethodOptions {
+		status = it.preflight(w, r)
+	} else {
+		status = it.request(w, r)
+	}
+
+	if status != 0 && status != safehttp.StatusNoContent {
+		if 400 <= status && status < 500 {
+			return w.ClientError(status)
+		} else {
+			return w.ServerError(status)
+		}
+	}
+
+	if origin != "" {
+		allowOrigin([]string{origin})
+		vary([]string{"Origin"})
+	}
+	if r.Header.Get("Cookie") != "" && it.AllowCredentials {
+		// TODO: handle other credentials than cookies:
+		// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Credentials
+		allowCredentials([]string{"true"})
+	}
+
+	if status == safehttp.StatusNoContent {
+		return w.NoContent()
+	}
+	return safehttp.Result{}
+}
+
+// preflight handles requests that have the method OPTIONS.
+func (it *Interceptor) preflight(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.StatusCode {
+	rh := r.Header
+	method := rh.Get("Access-Control-Request-Method")
+	if method == "" {
+		return safehttp.StatusForbidden
+	}
+	wh := w.Header()
+	allowMethods, err := wh.Claim("Access-Control-Allow-Methods")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	headers := rh.Get("Access-Control-Request-Headers")
+	if headers != "" {
+		for _, h := range strings.Split(headers, ", ") {
+			h = textproto.CanonicalMIMEHeaderKey(h)
+			if !it.allowedHeaders[h] && h != requiredHeader {
+				return safehttp.StatusForbidden
+			}
+		}
+	}
+	allowHeaders, err := wh.Claim("Access-Control-Allow-Headers")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	maxAge, err := wh.Claim("Access-Control-Max-Age")
+	if err != nil {
+		return safehttp.StatusInternalServerError
+	}
+
+	allowMethods([]string{method})
+	if headers != "" {
+		allowHeaders([]string{headers})
+	}
+	n := it.MaxAge
+	if n == 0 {
+		n = 5
+	}
+	maxAge([]string{strconv.Itoa(n)})
+	return safehttp.StatusNoContent
+}
+
+// request handles all requests that are not preflight requests.
+func (it *Interceptor) request(w safehttp.ResponseWriter, r *safehttp.IncomingRequest) safehttp.StatusCode {
+	h := r.Header
+	if h.Get(requiredHeader) != "1" {
+		return safehttp.StatusPreconditionFailed
```

Comment:
We *might* want to tell the client smth, like a message saying that header is required.

Reply:
True, but at the moment there is no way of sending error responses with messages. I see that your issue #127 would address this, but currently this is not possible.